integration/vite-hmr-hdr-test.ts

```diff
@@ -50,106 +47,145 @@ const indexRoute = `
   }
 `;
 
-test.describe("Vite HMR & HDR", () => {
-  templates.forEach(({ templateName, templateDisplayName }) => {
-    test.describe(templateDisplayName, () => {
-      test("vite dev", async ({ page, browserName, dev }) => {
-        let files: Files = async ({ port }) => ({
-          "vite.config.js": await viteConfig.basic({ port, templateName }),
-          "app/routes/_index.tsx": indexRoute,
-        });
-        let { cwd, port } = await dev(files, templateName);
-        await workflow({ templateName, page, browserName, cwd, port });
+templates.forEach((template) => {
+  const isRsc = template.name.includes("rsc");
+  const viteConfig = isRsc
+    ? tsx`
+      import { unstable_reactRouterRSC as reactRouterRSC } from '@react-router/dev/vite';
+      import rsc from '@vitejs/plugin-rsc';
+
+      export default {
+        plugins: [
+          reactRouterRSC(),
+          rsc(),
+        ],
+      };
+    `
+    : tsx`
+      import { reactRouter } from '@react-router/dev/vite';
+
+      export default {
+        plugins: [
+          reactRouter(),
+        ],
+      };
+    `;
+
+  test.describe(`${template.displayName} - HMR & HDR`, () => {
+    test.use({
+      template: template.name,
+      files: {
+        "vite.config.ts": viteConfig,
+        "app/routes/_index.tsx": indexRoute,
+      },
+    });
+
+    test("vite dev", async ({ page, edit, $ }) => {
+      const port = await getPort();
+      const url = `http://localhost:${port}`;
+
+      const dev = $(`pnpm dev --port ${port}`);
+      await Stream.match(dev.stdout, url);
+
+      await workflow({ templateName: template.name, page, edit, url });
+    });
+
+    test("express", async ({ page, edit, $ }) => {
+      await edit({
+        "server.mjs": isRsc ? Express.rsc() : Express.server(),
       });
 
-      test("express", async ({ page, browserName, customDev }) => {
-        let files: Files = async ({ port }) => ({
-          "vite.config.js": await viteConfig.basic({ port, templateName }),
-          "server.mjs": EXPRESS_SERVER({ port, templateName }),
```

Comment:
Previously, we set up Vite's `port` and `hmrPort` in `viteConfig.basic`. Turns out, when using middleware mode, only `hmrPort` matters. Also, rather than overly templatizing the Express `server.mjs`, I chose to use env vars for the inputs (port & HMR port) and separate functions for RSC and non-RSC modes (though I wouldn't be opposed to a wrapping funciton like `Express.server({ rsc: true })` )